pkg/streamingpromql/operator/ring_buffer.go

```diff
@@ -26,23 +26,50 @@ func (b *RingBuffer) DiscardPointsBefore(t int64) {
 	}
 }
 
-// Points returns slices of the points in this buffer.
+// UnsafePoints returns slices of the points in this buffer, including only points with timestamp less than or equal to maxT.
 // Either or both slice could be empty.
-// Callers must not modify the values in the returned slices.
+// Callers must not modify the values in the returned slices or return them to a pool.
+// Calling UnsafePoints is more efficient than calling CopyPoints, as CopyPoints will create a new slice and copy all
+// points into the slice, whereas UnsafePoints returns a view into the internal state of this buffer.
+// The returned slices are no longer valid if this buffer is modified (eg. a point is added, or the buffer is reset).
 //
 // FIXME: the fact we have to expose this is a bit gross, but the overhead of calling a function with ForEach is terrible.
 // Perhaps we can use range-over function iterators (https://go.dev/wiki/RangefuncExperiment) once this is not experimental?
-func (b *RingBuffer) Points() ([]promql.FPoint, []promql.FPoint) {
-	endOfTailSegment := b.firstIndex + b.size
+func (b *RingBuffer) UnsafePoints(maxT int64) ([]promql.FPoint, []promql.FPoint) {
```

Comment:
[nit] Could be easier to understand what it returns.
```suggestion
func (b *RingBuffer) UnsafePoints(maxT int64) (head []promql.FPoint, tail []promql.FPoint) {
```